specifyweb/frontend/js_src/lib/components/QueryBuilder/FieldFilter.tsx

```diff
@@ -161,7 +161,7 @@ export function QueryInputField({
             .map(({ parsed }) => (parsed as number | string).toString())
             .join(', ');
 
-      handleChange?.(parsed);
+      if (parsed) handleChange?.(parsed);
```

Comment:
this seems like it introduces a bug
if I had a search query in the field before, and now I remove it, the search query would still be present

Reply:
@maxxxxxdlp Do you mean: 
create stateFieldsJSONRef in FieldFilter.tsx 
like so: 
`  const stateFieldsJSONRef = React.useRef('');
  const stateFieldsJSON = JSON.stringify(state.fields);
  stateFieldsJSONRef.current = stateFieldsJSON;`
  
  And have a check in the useEffect: 
  `React.useEffect(() => {
    handleChange?.({
      fields: unParseQueryFields(state.baseTableName, state.fields),
      isDistinct: query.selectDistinct,
    });
    if (JSON.stringify(state.fields) !== stateFieldsJSONRef.current)
      saveRequired = true;
  }, [state, query.selectDistinct]);`
  ?

Reply:
generally yes

except the assignment to ref should only happen once, not on every render.
and use `Json` instead of `JSON` when naming (to follow strict camel case)

<details>
<summary>Code (look at this to compare after you tried implementing the solution)</summary>

  In Wrapped.tsx we currently have this:
  ```tsx
    const [state, dispatch] = React.useReducer(reducer, pendingState);
    React.useEffect(() => {
      dispatch({
        type: 'ResetStateAction',
        state: buildInitialState(),
      });
    }, [buildInitialState]);
  ```

  I would replace it with this:

  ```tsx
    const [state, dispatch] = React.useReducer(reducer, pendingState);
    const initialFields = React.useRef<string>('');
    React.useEffect(() => {
      const initialState = buildInitialState();
      dispatch({
        type: 'ResetStateAction',
        state: initialState,
      });
      initialFields.current = JSON.stringify(initialState.fields);
    }, [buildInitialState]);
  
    const saveRequired = React.useMemo(()=>
      initialFields.current !== JSON.stringify(state.fields)
    ,[state.fields]);
  ```

  Then can remove saveRequired from the state.

  But, this could cause performance problems (that JSON.stringify would be executed on every keystroke when typing a query field value - which could be slow if there are a lot of query fields and computer is not powerful, thus it can be optimized further by using throttle

  So you can replace this part:
  
  ```tsx
      const saveRequired = React.useMemo(()=>
        initialFields.current !== JSON.stringify(state.fields)
      ,[state.fields]);
  ```
  
  with this part:
  
  ```tsx
    const [saveRequried, setSaveRequired] = React.useState(false);
    const checkForChanges =
      throttle(() => {
        setSaveRequired(
          initialFields.current !== JSON.stringify(state.fields)
        );
      }, 200);
    React.useEffect(checkForChanges, [state.fields]); // or can make the call to checkForChanges() be part of the existing useEffect that calls handleChange, like in your example
  ```
  
  (optionally, if you are interested, you can read about how to analyze page performance to be able to inspect what optimization is necessary and when - https://developer.chrome.com/docs/devtools/performance/)

</details>